File: src/discord/commands/roll_stats.rb
# rubocop:disable Metrics/CyclomaticComplexity, Metrics/AbcSize, Metrics/ParameterLists
class Discord::Commands::RollStats
  def initialize(number_of_min: 0, number_of_max: 2, min_stat: 10, max_stat: 15, stat_baseline: 0, total_min: 70, total_max: 80)
    @number_of_min = number_of_min
    @number_of_max = number_of_max
    @min_stat = min_stat
    @max_stat = max_stat
    @stat_baseline = stat_baseline
    @total_min = total_min
    @total_max = total_max
    @stats = []
    @status = false
  end

  def perform
    return unless validate

    roll_stats until enough?

    stats
  end

  private

  attr_reader :number_of_min, :number_of_max, :max_stat, :min_stat, :stat_baseline, :total_min, :total_max
  attr_accessor :stats

  def validate
    return false if number_of_min + number_of_max > 6
    return false if number_of_min > 4 || number_of_max > 4
    return false if min_stat > max_stat
    return false if stat_baseline >= min_stat
    return false if stat_baseline > 8

    true
  end

  def roll_stats
    @stats = []

    @stats.push(roll_stat) until @stats.size == 6
  end

  def roll_stat
    values = [rand(1..6), rand(1..6), rand(1..6), rand(1..6)]
    top3 = values.sort { |x, y| y <=> x }.first(3)
    top3.inject(:+)
  end

  def enough?
    high_enough = 0
    low_enough = 0
    baseline_count = 0
    sum = 0

    stats.each do |stat|
      high_enough += 1 if stat > max_stat
      low_enough += 1 if stat < min_stat
      baseline_count += 1 if stat >= stat_baseline
      sum += stat
    end

    validate_stats(high_enough, low_enough, baseline_count, sum)
  end

  def validate_stats(high_enough, low_enough, baseline_count, sum)
    high_enough >= number_of_max && low_enough >= number_of_min && baseline_count == 6 && sum >= total_min && sum <= total_max
  end
end
# rubocop:enable Metrics/CyclomaticComplexity, Metrics/AbcSize, Metrics/ParameterLists

File: lib/dicer/roll_stats.ex
defmodule Dicer.RollStats do
  # Number of stats with less (or equal) than @min_stat value
  @number_of_min 0
  # Number of stats with more (or equal) @max_stat value
  @number_of_max 2
  # Min stat treshold
  @min_stat 10
  # Max stat treshold
  @max_stat 15
  # The lowest allowed stat
  @stat_baseline 0
  # The minimum sum of all stats
  @total_min 70
  # The maximum sum of all stats
  @total_max 80

  @spec call :: {:error, :max_iterations} | {:ok, [integer]}
  def call() do
    roll_stats(0)
  end

  # After 10,000 iterations just break the loop to save CPU
  defp roll_stats(10000) do
    {:error, :max_iterations}
  end

  defp roll_stats(tries) do
    stats = for(_i <- 1..6, do: roll_stat())

    if enough?(stats) do
      {:ok, stats}
    else
      roll_stats(tries + 1)
    end
  end

  defp roll_stat do
    [
      Enum.random(1..6),
      Enum.random(1..6),
      Enum.random(1..6),
      Enum.random(1..6)
    ]
    |> Enum.sort()
    |> Enum.take(3)
    |> Enum.sum()
  end

  defp enough?(stats) do
    Enum.reduce(
      stats,
      %{
        high_enough: 0,
        low_enough: 0,
        baseline_count: 0,
        sum: 0
      },
      fn stat, meta ->
        meta =
          if stat > @max_stat,
            do: increment_meta(meta, :high_enough),
            else: meta

        meta =
          if stat < @min_stat, do: increment_meta(meta, :low_enough), else: meta

        meta =
          if stat >= @stat_baseline,
            do: increment_meta(meta, :baseline_count),
            else: meta

        meta = increment_meta(meta, :sum, stat)

        meta
      end
    )
    |> validate_meta()
  end

  defp increment_meta(meta, key, value \\ 1) do
    Map.put(meta, key, meta[key] + value)
  end

  defp validate_meta(%{
         high_enough: high_enough,
         low_enough: low_enough,
         baseline_count: baseline_count,
         sum: sum
       }) do
    high_enough >= @number_of_max && low_enough >= @number_of_min &&
      baseline_count == 6 && sum >= @total_min && sum <= @total_max
  end
end
